group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2415,6 +2415,22 @@ private CoordinatorResult<Void, Record> classicGroupJoinExistingMember(
         return EMPTY_RESULT;
     }
 
+    /**
+     * An overload of {@link GroupMetadataManager#completeClassicGroupJoin(ClassicGroup)} used as
+     * timeout operation. It additionally looks up the group by the id and checks the group type.
+     * completeClassicGroupJoin will only be called if the group is CLASSIC.
+     */
+    private CoordinatorResult<Void, Record> completeClassicGroupJoin(String groupId) {
+        ClassicGroup group;
+        try {
+            group = getOrMaybeCreateClassicGroup(groupId, false);
+        } catch (UnknownMemberIdException | GroupIdNotFoundException exception) {
```

Comment:
I actually wonder whether raising `GroupIdNotFoundException` in `getOrMaybeCreateClassicGroup` is correct. It is likely not expected on the join-group and the sync-group paths. @jeffkbkim What do you think?

Reply:
The existing coordinator completes responses with UNKNOWN_MEMBER_ID instead of GROUP_ID_NOT_FOUND for join and sync. DeleteGroup and DeleteOffset requests are completed with GROUP_ID_NOT_FOUND if the group does not exist.

So just returning null and handling them separately for each API might be correct.

Reply:
This makes sense. My concern is more on the join-group/sync-group paths. We use `getOrMaybeCreateClassicGroup` there and don't seem to convert `GroupIdNotFoundException`. Do we need to?

Reply:
Filed https://issues.apache.org/jira/browse/KAFKA-16503.